flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorCoordinatorCheckpoints.java

```diff
@@ -58,7 +58,17 @@
 		final Collection<CompletableFuture<CoordinatorSnapshot>> individualSnapshots = new ArrayList<>(coordinators.size());
 
 		for (final OperatorCoordinatorCheckpointContext coordinator : coordinators) {
-			individualSnapshots.add(triggerCoordinatorCheckpoint(coordinator, checkpointId));
+			final CompletableFuture<CoordinatorSnapshot> checkpointFuture = triggerCoordinatorCheckpoint(coordinator, checkpointId);
+			coordinator.onCallTriggerCheckpoint(checkpointId);
+
+			individualSnapshots.add(checkpointFuture);
+			checkpointFuture.whenComplete((ignored, failure) -> {
+				if (failure != null) {
+					coordinator.abortCurrentTriggering();
+				} else {
+					coordinator.onCheckpointStateFutureComplete(checkpointId);
```

Comment:
If the `checkpointFuture` is completed by the coordinator executor before `whenComplete` is invoked here which shuts the valve, it seems possible that some events get into the valve after the snapshot is taken but before the valve is shut. That may result in inconsistent state between the TM and Coordinator.

Reply:
This sounded like a great catch to me initially. I have not tried to put this into a test to guard/fix this specific situation, but I found that I have a hard time to even come up with an implementation where this could even potentially happen.

Do you happen to have a code snippet that would produce the situation you where thinking of here?

Reply:
I don't have a code snippet in hand right now. But I can create one and run in a tight loop to show the race condition.

Reply:
I found a way to implement this test scenario: https://github.com/apache/flink/pull/12234/commits/6af459b5a07576f48a600484515ac30d72fcc4eb#diff-89f8c4f657c748242a72e9efdf5e7f97R285